Implementing callbacks p2

haneric00 opened the pull request:
Implemented the rest of the callbacks that the traceloop implementation had implemented
- behavior is basically the same from limited manual testing
- removed deprecated and proprietary attributes
- removed some unnecessary helper functions as a result of removing deprecated and proprietary attributes

haneric00 committed:
implemented on_llm_end and on_chat_model_start

- added a bunch of helpers (from openLLMetry implementation)
- identified a bunch of deprecated attributes and commented out the logic which corresponds with them

haneric00 committed:
implemented on_chain_start

this one was problematic because the openLLMetry implementation relies heavily on their proprietary attribute scheme, but with the help of bedrock we made the change

- used GPT generated alternative naming schemes to replace TraceLoop attributes because there seems to be no good official OTel alternative

haneric00 committed:
finished the callbacks (except agent because not included in the openLLMetry implementation)

TODOs: write the callbacks for agent

haneric00 committed:
addressing PR comments
- fixing typos
- deleting unecessary stuff

haneric00 committed:
final PR comment addressing:

- delete token
- delete workflow name
- delete entity path

haneric00 committed:
deleted a bunch of stuff
- deleted deprecated attributes
- deleted unused functions
- attempted to match naming convention of genAI spans according to OTel spec

haneric00 committed:
mid implementation of necessary fixes

- corrected naming
- in the middle of investigating where the context is going when the chain span starts

haneric00 committed:
about done with this PR

- behavior looks about the same as TraceLoop implementation

ezhang6811 commented on src/opentelemetry/instrumentation/langchain_v2/__init__.py:
```diff
@@ -16,6 +16,10 @@
 
 class LangChainInstrumentor(BaseInstrumentor):
+    @classmethod
+    def instrumentation_dependencies(cls) -> tuple:
```
use same return type as upstream

haneric00 replied:
changed to Collection[str]

ezhang6811 commented on src/opentelemetry/instrumentation/langchain_v2/callback_handler.py:
```diff
@@ -52,6 +43,12 @@ def _set_request_params(span, kwargs, span_holder: SpanHolder):
+        model = None
+
+    print(f"curr span holder: {span_holder}")
+    print(f"MODEL RIGHT HERE: {model}")
```
remove these logs before merging.

ezhang6811 commented on src/opentelemetry/instrumentation/langchain_v2/callback_handler.py:
```diff
@@ -78,9 +75,7 @@ def _set_request_params(span, kwargs, span_holder: SpanHolder):
+        tool_function = tool.get("function", tool)
         _set_span_attribute(
             span,
             f"{Span_Attributes.GEN_AI_TOOL_NAME}.{i}",
```
Otel semantic conventions doesn't allow `gen_ai.tool.name` to be an enumerated list. I'm not sure why upstream has this logic here, but we should only be instrumenting tool name from the tool call callbacks, which should each create one span with one tool name + type.

ezhang6811 commented on src/opentelemetry/instrumentation/langchain_v2/callback_handler.py:
```diff
@@ -106,43 +99,36 @@ def _set_span_attribute(span: Span, name: str, value: AttributeValue):
+        ):   
+            _set_span_attribute(span, f"{Span_Attributes.GEN_AI_TOOL_NAME}.{i}", function.get("name"))
+            _set_span_attribute(span, f"{Span_Attributes.GEN_AI_TOOL_DESCRIPTION}.{i}", function.get("description"))
+            _set_span_attribute(span, f"{Span_Attributes.GEN_AI_TOOL_TYPE}.{i}", json.dumps(function.get("parameters")))
```
The chat callback should not be using tool call attributes. You might need to set up a chat model example here to actually see which attributes in kwargs or serialized can map to Otel attributes. could remove for now and save for a future PR.

ezhang6811 commented on src/opentelemetry/instrumentation/langchain_v2/callback_handler.py:
```diff
@@ -260,90 +250,224 @@ def _get_name_from_callback(
+
+        model_name = None
+        if response.llm_output is not None:
+            span.update_name(f"{model_name}") if model_name else span.update_name(span.name)
```
why do we need this line?

ezhang6811 commented on src/opentelemetry/instrumentation/langchain_v2/callback_handler.py:
```diff
@@ -260,90 +250,224 @@ def _get_name_from_callback(
+        name = self._get_name_from_callback(serialized, **kwargs)
+        kind = SpanKind.INTERNAL
+
+        span = self._create_task_span(
```
This function in OpenLLMetry is only used for traceloop specific attributes, you should be able to remove it and directly create the span here.

haneric00 replied:
do you mean _get_name_from_callback or _create_task_span? both functions from a quick glance look pretty useful/universal to me

ezhang6811 commented on src/opentelemetry/instrumentation/langchain_v2/callback_handler.py:
```diff
@@ -260,90 +250,224 @@ def _get_name_from_callback(
+            kind,
+            metadata,
+        )
+        should_trace = context_api.get_value("override_enable_content_tracing") or True
```
remove these flags for suppressing/overriding tracing for now.

ezhang6811 commented on src/opentelemetry/instrumentation/langchain_v2/callback_handler.py:
```diff
@@ -260,90 +250,224 @@ def _get_name_from_callback(
+        if should_trace:
+            _set_span_attribute(
+                span,
+                "app.entity.input", # gpt generated alternative naming scheme
```
we don't need this attribute at all if there isn't an Otel equivalent. Should look for an attribute that can be mapped to `gen_ai.conversation.id`.

ezhang6811 commented on src/opentelemetry/instrumentation/langchain_v2/callback_handler.py:
```diff
@@ -260,90 +250,224 @@ def _get_name_from_callback(
+        if should_trace:
+            _set_span_attribute(
+                span,
+                "app.entity.output",
```
Instead of this, you can hardcode a generic attribute for now like `chain.output` to store the outputs object.

ezhang6811 commented on src/opentelemetry/instrumentation/langchain_v2/callback_handler.py:
```diff
@@ -260,90 +250,224 @@ def _get_name_from_callback(
+            SpanKind.INTERNAL,
+        )
+        
+        should_trace = context_api.get_value("override_enable_content_tracing") or True
```
This is where we should be setting tool id (if it exists), name, type, and description. Again, you probably need to do some input logging for a tool call in your sample app to find where these attributes reside.

ezhang6811 commented on src/opentelemetry/instrumentation/langchain_v2/callback_handler.py:
```diff
@@ -260,90 +250,224 @@ def _get_name_from_callback(
+        if should_trace:
+            _set_span_attribute(
+                span,
+                "app.entity.output",
```
similar to chain.output, just use a generic `tool.output` for now.

ezhang6811 commented on src/opentelemetry/instrumentation/langchain_v2/span_attributes.py:
```diff
@@ -33,11 +33,12 @@ class Span_Attributes:
     GEN_AI_TOOL_NAME = "gen_ai.tool.name"
     GEN_AI_TOOL_DESCRIPTION = "gen_ai.tool.description"
     GEN_AI_TOOL_TYPE = "gen_ai.tool.type"
+    GEN_AI_TOKEN_TYPE = "gen_ai.token.type"
```
remove if not used. seems like only for metrics

ezhang6811 commented on src/opentelemetry/instrumentation/langchain_v2/utils.py:
```diff
@@ -1,8 +1,41 @@
+from pydantic import BaseModel
+
+
+class CallbackFilteredJSONEncoder(json.JSONEncoder):
```
Can we remove this if we delete all the instances of "app.entity.output" that just dump the entire json into an attribute?

haneric00 replied:
yes

ezhang6811 replied:
bump once you remove all the json.dumps attributes above.

ezhang6811 commented on src/opentelemetry/instrumentation/langchain_v2/callback_handler.py:
```diff
@@ -260,90 +250,224 @@ def _get_name_from_callback(
-        pass
-    
-    
+    def on_agent_error(self, error, *, run_id, parent_run_id, **kwargs):
```
save this change for the agent PR

ezhang6811 commented on src/opentelemetry/instrumentation/langchain_v2/callback_handler.py:
```diff
@@ -260,90 +250,224 @@ def _get_name_from_callback(
+
+    
+    @dont_throw
+    def on_chat_model_start(self, serialized, messages, *, run_id, tags, parent_run_id, metadata, **kwargs):
```
can you use the expanded function signature for every function like above? i.e. format each parameter to be its own line and specify the parameter types.

haneric00 replied:
done

ezhang6811 reviewed (commented):
I'm a little dubious of the use of SpanHolder. It might be better if we just manage our own mapping of run_id's directly to the spans.

haneric00 committed:
mid PR comment addressing time to commit
- deleting unnecessary features/leftovers from TraceLoop implementation which we are not using
- deleting unused attributes
- lots more deleting of comments and such

haneric00 committed:
quick fix: removed a change to agent callback: we save these for a future PR

haneric00 committed:
deleted some unnecessary code (anecdotal manual testing shows they did literally nothing)

haneric00 committed:
deleted some more stuff + added 2 attributes to on_tool_start

haneric00 committed:
not sure if my tools are tracing correctly

TODOs: validate correctness of tool tracing and delete prints

haneric00 committed:
ready to PR check for merge:

- deleted some unnecessary helpers/simplified helper function pipeline

- deleted prints

goal to accomplish:
- fix tool callbacks (no more weird indexing)

ezhang6811 commented on src/opentelemetry/instrumentation/langchain_v2/callback_handler.py:
```diff
@@ -260,90 +188,318 @@ def _get_name_from_callback(
+            return
+
+        model_id = None
+        if "invocation_params" in kwargs and "model_id" in kwargs["invocation_params"]:
```
Why don't we have to check for the model_id as "model" or "model_name" like we do [here](https://github.com/haneric00/opentelemetry-langchain-instrumentation-v2/blob/59959aa936fbba920dad4ee8bed1ff36149fa101/src/opentelemetry/instrumentation/langchain_v2/callback_handler.py#L38)?

ezhang6811 commented on src/opentelemetry/instrumentation/langchain_v2/callback_handler.py:
```diff
@@ -260,90 +188,318 @@ def _get_name_from_callback(
+            if model_name is not None:
+                _set_span_attribute(span, Span_Attributes.GEN_AI_RESPONSE_MODEL, model_name)
+
+                if self.span_mapping[run_id].request_model is None:
```
Nit: Don't think this is needed/would ever happen.

ezhang6811 commented on src/opentelemetry/instrumentation/langchain_v2/callback_handler.py:
```diff
@@ -260,90 +188,318 @@ def _get_name_from_callback(
+            completion_tokens = (
+                token_usage.get("completion_tokens")
+                or token_usage.get("generated_token_count")
+                or token_usage.get("output_tokens")
```
note: you should eventually test this with every model family that Langchain supports - there could be other names for the input/output tokens not accounted for here. This is a requirement for us to say we support instrumentation for all models, or worst case a specific subset if there are unreasonably many different naming conventions.

ezhang6811 commented on src/opentelemetry/instrumentation/langchain_v2/callback_handler.py:
```diff
@@ -260,90 +188,318 @@ def _get_name_from_callback(
+                    "metadata": metadata,
+                    "kwargs": kwargs,
+                },
+                cls=CallbackFilteredJSONEncoder,
```
I still think it would be better to just set "chain.input" to the string form of "inputs". Same with "chain.output" and "outputs", etc. below. The "tags" and "metadata" attributes aren't very useful to customers, and anything useful in "kwargs" we should've already converted to our naming conventions.

ezhang6811 commented on src/opentelemetry/instrumentation/langchain_v2/callback_handler.py:
```diff
@@ -260,90 +188,318 @@ def _get_name_from_callback(
+                },
+                cls=CallbackFilteredJSONEncoder,
+            ),
+        ) ######### lowkey useless span attribute
```
Just set directly to whatever is in "input_str", the rest can be removed.

ezhang6811 commented on src/opentelemetry/instrumentation/langchain_v2/callback_handler.py:
```diff
@@ -260,90 +188,318 @@ def _get_name_from_callback(
+                {"output": output, "kwargs": kwargs},
+                cls=CallbackFilteredJSONEncoder,
+            ), 
+        ) ######## lowkey useless tool output
```
same as above, just set to the value of "output". Also you can specify output's type as str instead of Any. Don't forget to remove the comment.

haneric00 committed:
addressed PR comments:

- deleted unnecessary json.dumps attributes
- deleted the helper function facilitating previous
- replaced previous with normal string input/outputs
- modified attribute handling for model id (ONLY for bedrock purposes, checked bedrock implementation and they only have 2 param names)

haneric00 committed:
delete unused helper in utils

haneric00 committed:
changed default value for model id

haneric00 committed:
delete unused import

haneric00 committed:
Merge branch 'main' into implementing-callbacks-p2
